File: 11/01.py
input = open('input.txt').read().strip().split(',')

coords = [0, 0]

directions = {
    's': [0, 1],
    'n': [0, -1],
    'e': [1, 0],
    'w': [-1, 0],
}

for i, move in enumerate(input):
    odd = abs(coords[0]) % 2
    if len(move) == 2:
        d = [x[0] + x[1] for x in zip(directions[move[1]], directions[move[0]])]
        if not odd and move[0] == 's':
            d = [d[0], d[1] - 1]
        if odd and move[0] == 'n':
            d = [d[0], d[1] + 1]
    else:
        d = directions[move]
    coords = [coords[0] + d[0], coords[1] + d[1]]

moves = 0
while coords != [0, 0]:
    odd = abs(coords[0]) % 2
    moves += 1
    if coords[1] != 0:
        coords = [coords[0] + 1, coords[1] - 1 if not odd else coords[1]]
    else:
        coords = [coords[0] + 1, coords[1]]
assert moves == 773
print(moves)

File: 20/01.rb
class Particle
  attr_reader :position, :velocity, :acceleration, :id

  def initialize(id, position, velocity, acceleration)
    @id = id
    @position = position
    @velocity = velocity
    @acceleration = acceleration
  end

  def dist
    [@position.x.abs, @position.y.abs, @position.z.abs].sum
  end

  def move!
    @velocity.x += @acceleration.x
    @velocity.y += @acceleration.y
    @velocity.z += @acceleration.z
    @position.x += @velocity.x
    @position.y += @velocity.y
    @position.z += @velocity.z
  end
end

class Coord
  attr_accessor :x, :y, :z
  def initialize(x, y, z)
    @x = x
    @y = y
    @z = z
  end
end

particles = File.foreach('input.txt', "\n").map.with_index do |line, i|
  coords = line.scan(/[a-z]=<(.*?)>/).map do |xyz|
    Coord.new(*xyz[0].split(',').map(&:to_i))
  end
  Particle.new(i, *coords)
end

1000.times do
  particles.each(&:move!)
end

closest = particles.reduce(nil) do |accum, obj|
  accum = obj if accum.nil?
  obj.dist < accum.dist ? obj : accum
end

puts(closest.id)
raise 'wrong answer' if closest.id != 161
